pkg/networking/session.go

```diff
@@ -0,0 +1,411 @@
+		s.logger.Debug("Data written into send channel")
+	case <-s.ctx.Done():
+		s.logger.Debug("Session shutdown while sending data")
+		return ErrSessionShutdown
+	case <-timer.C:
+		s.logger.Debug("Connection write timeout while sending data")
+		return ErrConnectionWriteTimeout
+	}
+
+	dataCopy := func() {
+		if data == nil {
+			return // A nil data is ignored.
+		}
+
+		// In the event of session shutdown or connection write timeout, we need to prevent `send` from reading
+		// the body buffer after returning from this function since the caller may re-use the underlying array.
+		ready.mu.Lock()
+		defer ready.mu.Unlock()
+
+		if ready.data == nil {
+			return // data was already copied in `send`.
+		}
+		newData := make([]byte, len(data))
+		copy(newData, data)
+		ready.data = newData
+	}
+
+	select {
+	case err := <-errCh:
+		s.logger.Debug("Data sent", "error", err)
+		return err
+	case <-s.ctx.Done():
+		dataCopy()
+		s.logger.Debug("Session shutdown while waiting send error")
+		return ErrSessionShutdown
+	case <-timer.C:
+		dataCopy()
+		s.logger.Debug("Connection write timeout while waiting send error")
+		return ErrConnectionWriteTimeout
+	}
+}
+
+// sendLoop is a long-running goroutine that sends data to the connection.
+func (s *Session) sendLoop() error {
+	var dataBuf bytes.Buffer
+	for {
+		dataBuf.Reset()
+
+		select {
+		case <-s.ctx.Done():
+			s.logger.Debug("Exiting connection send loop")
+			return s.ctx.Err()
+
+		case packet := <-s.sendCh:
+			s.logger.Debug("Sending data to connection",
+				"data", base64.StdEncoding.EncodeToString(packet.data))
+			packet.mu.Lock()
+			if packet.data != nil {
+				// Copy the data into the buffer to avoid holding a mutex lock during the writing.
+				_, err := dataBuf.Write(packet.data)
+				if err != nil {
+					packet.data = nil
+					packet.mu.Unlock()
+					s.logger.Error("Failed to copy data into buffer", "error", err)
+					s.asyncSendErr(packet.err, err)
+					return err // TODO: Do we need to return here?
+				}
+				s.logger.Debug("Data copied into buffer")
+				packet.data = nil
+			}
+			packet.mu.Unlock()
+
+			if dataBuf.Len() > 0 {
+				s.logger.Debug("Writing data into connection", "len", len(dataBuf.Bytes()))
+				_, err := s.conn.Write(dataBuf.Bytes()) // TODO: We are locking here, because no timeout set on connection itself.
+				if err != nil {
+					s.logger.Error("Failed to write data into connection", "error", err)
+					s.asyncSendErr(packet.err, err)
+					return err
+				}
+				s.logger.Debug("Data written into connection")
+			}
+
+			// No error, successful send
+			s.asyncSendErr(packet.err, nil)
+		}
+	}
+}
+
+// receiveLoop continues to receive data until a fatal error is encountered or underlying connection is closed.
+// Receive loop works after handshake and accepts only length-prepended messages.
+func (s *Session) receiveLoop() error {
+	s.establishedLock.Lock() // Prevents from running multiple receiveLoops.
+	defer s.establishedLock.Unlock()
```

Comment:
There should be only one `receiveLoop` running. And `receiveLoop` can be started only from `newSession` function.
Actually pairs like `establishedLock`/`established` and `shutdownLock`/`shutdown` can be replaced by `sync.Once`. I don't see why we should throw error instead of silently ignore an attempt to start another `receiveLoop`. 
For now, I'll replace the lock/bool pairs with `sync.Once`.

Reply:
I think the method should return an error in cases of incorrect usage. Running `receiveLoop` twice is incorrect usage.

Reply:
Once doesn't work for establishedLock, I did like you suggested but without returning an error on the attempt to start second receiveLoop.

Reply:
> I think the method should return an error in cases of incorrect usage. Running `receiveLoop` twice is incorrect usage.

It's a private function, no way to use it incorrectly outside the package. Better panic because it's the developer error then. But ignoring also works well.

Reply:
Okay 👌